openlibrary/core/bookshelves_events.py

```diff
@@ -84,7 +84,46 @@ def select_by_book_user_and_type(cls, username, work_id, edition_id, event_type)
             event_type=$event_type
         """
 
-        return list(db.select(cls.TABLENAME, where=where, vars=data))
+        return list(oldb.select(cls.TABLENAME, where=where, vars=data))
+
+    @classmethod
+    def select_by_user_type_and_year(cls, username, event_type, year):
+        oldb = db.get_db()
+
+        data = {
+            'username': username,
+            'event_type': event_type,
+            'event_date': f'{year}%',
+        }
+
+        where = """
+            username=$username AND
+            event_type=$event_type AND
+            event_date LIKE $event_date
+        """
+
+        return list(oldb.select(cls.TABLENAME, where=where, vars=data))
+
+    @classmethod
+    def select_distinct_by_user_type_and_year(cls, username, event_type, year):
+        """Returns a list of the most recent check-in events, with no repeating
+        work IDs.  Useful for calculating one's yearly reading goal progress.
+        """
+        oldb = db.get_db()
+
+        data = {
+            'username': username,
+            'event_type': event_type,
+            'event_date': f'{year}%',
+        }
+        query = (
+            f"select distinct on (work_id) work_id, * from {cls.TABLENAME} "
+            "where username=$username and event_type=$event_type and "
+            "event_date LIKE $event_date "
+            "order by work_id, updated desc"
+        )
+
+        return list(oldb.query(query, vars=data))
```

Comment:
Do you think there may be an opportunity to merge these 2 functions to make them a bit more concise & flexible?

Instead of `select_by_user_type_and_year` and `select_distinct_by_user_type_and_year`, perhaps `distinct` could be a boolean flag?

Furthermore, if we made this merge, I wonder if we may be able to simplify the name as `get_users_events` (consistent with naming style in https://github.com/internetarchive/openlibrary/blob/master/openlibrary/core/bookshelves.py#L185)

The definition could look like:

```
def get_users_events(self, username, type=None, year=None, distinct=False)
```

This would allow us to get all events by username, filter optionally by a type, get all years by default (or filter by year), with the option to count distinct.



```suggestion
    @classmethod
    def select_by_user_type_and_year(cls, username, event_type, year):
        oldb = db.get_db()

        data = {
            'username': username,
            'event_type': event_type,
            'event_date': f'{year}%',
        }

        where = """
            username=$username AND
            event_type=$event_type AND
            event_date LIKE $event_date
        """

        return list(oldb.select(cls.TABLENAME, where=where, vars=data))

    @classmethod
    def select_distinct_by_user_type_and_year(cls, username, event_type, year):
        """Returns a list of the most recent check-in events, with no repeating
        work IDs.  Useful for calculating one's yearly reading goal progress.
        """
        oldb = db.get_db()

        data = {
            'username': username,
            'event_type': event_type,
            'event_date': f'{year}%',
        }
        query = (
            f"select distinct on (work_id) work_id, * from {cls.TABLENAME} "
            "where username=$username and event_type=$event_type and "
            "event_date LIKE $event_date "
            "order by work_id, updated desc"
        )

        return list(oldb.query(query, vars=data))
```